x/emissions/abci.go

```diff
@@ -35,71 +37,92 @@ func BeginBlocker(ctx sdk.Context, keeper keeper.Keeper, stakingKeeper types.Sta
-}
-func GetBondFactor(ctx sdk.Context, stakingKeeper types.StakingKeeper, keeper keeper.Keeper) sdk.Dec {
-	targetBondRatio := sdk.MustNewDecFromStr(keeper.GetParams(ctx).TargetBondRatio)
-	maxBondFactor := sdk.MustNewDecFromStr(keeper.GetParams(ctx).MaxBondFactor)
-	minBondFactor := sdk.MustNewDecFromStr(keeper.GetParams(ctx).MinBondFactor)
+func DistributeObserverRewards(ctx sdk.Context, amount sdkmath.Int, keeper keeper.Keeper) error {
 
-	currentBondedRatio := stakingKeeper.BondedRatio(ctx)
-	// Bond factor ranges between minBondFactor (0.75) to maxBondFactor (1.25)
-	if currentBondedRatio.IsZero() {
-		return sdk.ZeroDec()
+	rewardsDistributer := map[string]int64{}
+	totalRewardsUnits := int64(0)
+	err := keeper.GetBankKeeper().SendCoinsFromModuleToModule(ctx, types.ModuleName, types.UndistributedObserverRewardsPool, sdk.NewCoins(sdk.NewCoin(config.BaseDenom, amount)))
+	if err != nil {
+		return err
 	}
-	bondFactor := targetBondRatio.Quo(currentBondedRatio)
-	if bondFactor.GT(maxBondFactor) {
-		return maxBondFactor
+	ballotIdentifiers := keeper.GetObserverKeeper().GetMaturedBallotList(ctx)
+	// do not distribute rewards if no ballots are matured, the rewards can accumulate in the undistributed pool
+	if len(ballotIdentifiers) == 0 {
+		return nil
 	}
-	if bondFactor.LT(minBondFactor) {
-		return minBondFactor
+	for _, ballotIdentifier := range ballotIdentifiers {
+		ballot, found := keeper.GetObserverKeeper().GetBallot(ctx, ballotIdentifier)
+		if !found {
+			continue
+		}
+		totalRewardsUnits = totalRewardsUnits + ballot.BuildRewardsDistribution(rewardsDistributer)
+	}
+	rewardPerUnit := sdkmath.ZeroInt()
+	if totalRewardsUnits > 0 && amount.IsPositive() {
+		rewardPerUnit = amount.Quo(sdk.NewInt(totalRewardsUnits))
 	}
-	return bondFactor
-}
-
-func GetDurationFactor(ctx sdk.Context, keeper keeper.Keeper) sdk.Dec {
-	avgBlockTime := sdk.MustNewDecFromStr(keeper.GetParams(ctx).AvgBlockTime)
-	NumberOfBlocksInAMonth := sdk.NewDec(types.SecsInMonth).Quo(avgBlockTime)
-	monthFactor := sdk.NewDec(ctx.BlockHeight()).Quo(NumberOfBlocksInAMonth)
-	logValueDec := sdk.MustNewDecFromStr(keeper.GetParams(ctx).DurationFactorConstant)
-	// month * log(1 + 0.02 / 12)
-	fractionNumerator := monthFactor.Mul(logValueDec)
-	// (month * log(1 + 0.02 / 12) ) + 1
-	fractionDenominator := fractionNumerator.Add(sdk.OneDec())
 
-	// (month * log(1 + 0.02 / 12)) / (month * log(1 + 0.02 / 12) ) + 1
-	if fractionDenominator.IsZero() {
-		return sdk.OneDec()
+	sortedKeys := make([]string, 0, len(rewardsDistributer))
+	for k := range rewardsDistributer {
+		sortedKeys = append(sortedKeys, k)
 	}
-	if fractionNumerator.IsZero() {
-		return sdk.ZeroDec()
+	sort.Strings(sortedKeys)
+
+	var finalDistributionList []*types.ObserverEmission
+	for _, key := range sortedKeys {
+		observerAddress, err := sdk.AccAddressFromBech32(key)
+		if err != nil {
+			continue
```

Comment:
Thought?

Reply:
I already added a log line here 
https://github.com/zeta-chain/zeta-node/blob/0d2dce7a526870847899c267ca65989d06d40618/x/emissions/abci.go#L89-L89

Don't think we need to anything other than the log line though , since the chances of this error happening are impossible , as we do not have any way to add observers to the network , other than an upgrade

Reply:
Oh ok sorry, wasn't synchronized on my window 

> since the chances of this error happening are impossible

I think it's a valid reason to have a log because it would mean something is very wrong